core/tests/puppeteer-acceptance-tests/user-utilities/exploration-creator-utils.ts

```diff
@@ -13,146 +13,623 @@
+   * matches the expected category.
+   */
+  async expectSelectedCategoryToBe(expectedCategory: string): Promise<void> {
+    await this.page.waitForSelector('.mat-select-value');
+    const selectedCategory = await this.page.evaluate(() => {
+      return (
+        document.querySelector('.mat-select-value') as HTMLElement
+      ).innerText.trim();
+    });
+    if (selectedCategory === expectedCategory) {
+      showMessage(
+        `The category ${selectedCategory} is same as expectedCategory.`
+      );
+    } else {
+      throw new Error('Category is not correct.');
+    }
+  }
+
+  async selectLanguage(language: string): Promise<void> {
+    /**
+     * Language dropdown was visible, but when we click on that then options
+     * appear at wrong place which requires to click more than one time also it
+     * added some type of flake.So, to avoid that we are scrolling the page.
+     * We can use 300 - 500px to move the language dropdown to the upper
+     * part of the page.
+     */
+    await this.page.evaluate(() => {
+      window.scrollTo(0, 350);
+    });
+
+    await this.clickOn(languageUpdateDropDown);
+    await this.clickOn(language);
+  }
+
+  /**
+   *  This function verifies that the selected language matches the
+   *  expected language.
+   */
+  async expectSelectedLanguageToBe(expectedLanguage: string): Promise<void> {
+    await this.page.waitForSelector(languageUpdateDropDown);
+    const languageDropdown = await this.page.$(languageUpdateDropDown);
+    if (!languageDropdown) {
+      throw new Error('Category dropdown not found.');
+    }
+    await languageDropdown.click();
+
+    const selectedLanguage = await this.page.evaluate(() => {
+      const matOption = document.querySelector(
+        '.mat-option.mat-selected'
+      ) as HTMLElement;
+      if (!matOption) {
+        throw new Error('Selected language option not found.');
+      }
+      return matOption.innerText.trim();
+    });
+
+    if (selectedLanguage.includes(expectedLanguage)) {
+      showMessage(
+        `The language ${selectedLanguage} contains the expected language.`
+      );
+    } else {
+      throw new Error('Language is not correct.');
+    }
+    await this.page.keyboard.press('Enter');
+  }
+
+  async addTags(tagNames: string[]): Promise<void> {
+    for (let i = 0; i < tagNames.length; i++) {
+      await this.clickOn(addTagsInputBox);
+      await this.type(addTagsInputBox, tagNames[i].toLowerCase());
+      await this.page.keyboard.press('Tab');
+    }
+  }
+
+  async expectTagsToMatch(expectedTags: string[]): Promise<void> {
+    /**
+     * Converting tags to lowercase because while accessing added tags with
+     * `document.querySelectorAll('mat-chip')` we get tags converted in lowercase.
+     */
+    const lowercaseExpectedTags = expectedTags.map(tag => tag.toLowerCase());
+    await this.page.waitForSelector('mat-chip-list');
+    const addedTags = await this.page.evaluate(() => {
```

Comment:
This should be something like observedTags -- remember, this function checks the current state and should not have reference to previous actions.

Reply:
done